engine/table/src/main/java/io/deephaven/engine/table/impl/TimeTable.java

```diff
@@ -75,33 +99,29 @@ private void refresh(final boolean notifyListeners) {
         entry.onUpdateStart();
         try {
             final DateTime dateTime = timeProvider.currentTime();
-            DateTime currentBinnedTime = new DateTime(
-                    LongNumericPrimitives.lowerBin(dateTime.getNanos() - binOffset, period) + binOffset);
             long rangeStart = lastIndex + 1;
-            if (lastTime == null) {
+            if (columnSource.baseTime == null) {
                 lastIndex = 0;
-                dateTimeArraySource.ensureCapacity(lastIndex + 1);
-                dateTimeArraySource.set(lastIndex, lastTime = currentBinnedTime);
-                getRowSet().writableCast().insert(lastIndex);
-            } else
-                while (currentBinnedTime.compareTo(lastTime) > 0) {
-                    lastTime = DateTimeUtils.plus(lastTime, period);
-                    lastIndex++;
-                    dateTimeArraySource.ensureCapacity(lastIndex + 1);
-                    dateTimeArraySource.set(lastIndex, lastTime);
-                }
+                columnSource.baseTime = new DateTime(
+                        LongNumericPrimitives.lowerBin(dateTime.getNanos(), columnSource.period));
+            } else {
+                lastIndex = Math.max(lastIndex,
+                        DateTimeUtils.minus(dateTime, columnSource.baseTime) / columnSource.period);
+            }
+
             if (rangeStart <= lastIndex) {
                 // If we have a period longer than 10s, print out that the timetable has been updated. This can be
                 // useful when analyzing what's gone wrong in the logs. It is capped at periods of 5s, so we don't
                 // end up with too much log spam for short interval time tables. 5s is not so coincidentally the period
                 // of the Jvm Heap: messages.
-                if (period >= 5_000_000_000L) {
-                    log.info().append("TimeTable updated to ").append(lastTime.toString()).endl();
+                if (columnSource.period >= 5_000_000_000L) {
+                    log.info().append("TimeTable updated to ").append(columnSource.get(lastIndex).toString()).endl();
                 }
                 final RowSet range = RowSetFactory.fromRange(rangeStart, lastIndex);
-                getRowSet().writableCast().insert(range);
+                final RowSet removedRange = isStreaming ? getRowSet().copy() : RowSetFactory.empty();
+                getRowSet().writableCast().update(range, removedRange);
```

Comment:
I would consider just doing a `removeRange()` (if stream), and always doing an `insertRange()`. The outcome should be the same, but this way you tell the `RowSet` what you already know, rather than have it determine everything itself with `instanceof SingleRange` checks and similar.